File: sample_app_dagger/src/main/java/com/chumarin/stanislav/sample_app_dagger/login/LoginActivity.java
package com.chumarin.stanislav.sample_app_dagger.login;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.chumarin.stanislav.sample_app_dagger.MainActivity;
import com.chumarin.stanislav.sample_app_dagger.R;
import com.chumarin.stanislav.sample_app_dagger.di.ViewModels;

import dagger.Lazy;
import dagger.internal.DoubleCheck;
import io.reactivex.disposables.Disposable;

public class LoginActivity extends AppCompatActivity {

    private EditText userIdView;
    private Button loginButton;
    private View progressView;


    private Lazy<LoginViewModel> lazyViewModel = DoubleCheck.lazy(() -> ViewModels.create(this, LoginViewModel.class));

    private Disposable disposable;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        userIdView = findViewById(R.id.user_id);
        loginButton = findViewById(R.id.login);
        progressView = findViewById(R.id.loading);

        loginButton.setOnClickListener(view -> {
            String userId = userIdView.getText().toString().trim();
            if (!userId.isEmpty()) {
                lazyViewModel.get().onLoginButtonClicked(userId);
                hideKeyboard(userIdView);
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        disposable = lazyViewModel.get().observeState()
                .subscribe(state -> {
                    if (state instanceof LoginViewModel.State.Idle) {
                        loginButton.setEnabled(true);
                        progressView.setVisibility(View.GONE);
                    } else if (state instanceof LoginViewModel.State.Progress) {
                        loginButton.setEnabled(false);
                        progressView.setVisibility(View.VISIBLE);
                    } else if (state instanceof LoginViewModel.State.Success) {
                        LoginViewModel.State.Success success = (LoginViewModel.State.Success) state;
                        startActivity(MainActivity.createIntent(this, success.getUserId()));
                        finish();
                    }
                });
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (disposable != null) {
            disposable.dispose();
            disposable = null;
        }
    }

    private void hideKeyboard(@NonNull View view) {
        InputMethodManager inputMethodManager = (InputMethodManager) view.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

}

File: sample_app_koin/src/main/kotlin/vit/khudenko/android/sessiontracker/sample/koin/login/LoginActivity.kt
package vit.khudenko.android.sessiontracker.sample.koin.login

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel
import vit.khudenko.android.sessiontracker.sample.koin.EXTRA_CURRENT_SESSION_ID
import vit.khudenko.android.sessiontracker.sample.koin.MainActivity
import vit.khudenko.android.sessiontracker.sample.koin.R

class LoginActivity : AppCompatActivity() {

    private lateinit var userIdView: EditText
    private lateinit var loginButton: Button
    private lateinit var progressView: View

    private val viewModel: LoginViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        userIdView = findViewById(R.id.user_id)
        loginButton = findViewById(R.id.login)
        progressView = findViewById(R.id.loading)

        loginButton.setOnClickListener {
            val userId = userIdView.text.toString().trim()
            if (userId.isNotBlank()) {
                viewModel.onLoginButtonClicked(userId)
                hideKeyboard(userIdView)
            }
        }

        lifecycleScope.launch {
            viewModel.stateFlow()
                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
                .collect {
                    when (it) {
                        LoginViewModel.State.Idle -> {
                            loginButton.isEnabled = true
                            progressView.visibility= View.GONE
                        }
                        LoginViewModel.State.Progress -> {
                            loginButton.isEnabled = false
                            progressView.visibility= View.VISIBLE
                        }
                        is LoginViewModel.State.Success -> {
                            val intent = Intent(this@LoginActivity, MainActivity::class.java)
                            intent.putExtra(EXTRA_CURRENT_SESSION_ID, it.userId)
                            startActivity(intent)
                            finish()
                        }
                    }
                }
        }
    }

    private fun hideKeyboard(editText: EditText) {
        val inputMethodManager = editText.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(editText.windowToken, 0)
    }
}
